synapse/types.py

```diff
@@ -743,6 +744,26 @@ class ReadReceipt:
     data: JsonDict
 
 
+@attr.s(slots=True, frozen=True, auto_attribs=True)
+class DeviceLists:
+    """
+    Attributes:
+        changed: user_ids whose devices may have changed
+        left: user_ids whose devices we no longer track
+    """
+
+    # We need to use a factory here, otherwise `set` is not evaluated at
+    # object instantiation, but instead at class definition instantiation.
+    # The latter happening only once, thus always giving you the same sets
+    # across multiple DeviceLists instances.
+    # Also see: don't define mutable default arguments.
```

Comment:
We do, though why we do it isn't documented anywhere else, and I did initially struggle without using a factory.

It's not only relevant to this bit of code, though lacking anywhere else good to put it, leaving it may still help someone?

Reply:
I found this to be a fundamental property of using `attr`, but if you find it useful than that's fine with me.